### PR TITLE
Issue/6512 error exception cases

### PR DESCRIPTION
**Pull request by theck13**
#### Note
This should be merged ***after*** https://github.com/wordpress-mobile/WordPress-Android/pull/6671, https://github.com/wordpress-mobile/WordPress-Android/pull/6672, and https://github.com/wordpress-mobile/WordPress-Android/pull/6673.

### Fix
Add error/exception cases for failed Google login to detect when to display the error screen as described in https://github.com/wordpress-mobile/WordPress-Android/issues/6512.

### Test
The Google documentation has a list of status codes returned when an error is encountered during the login process.  The statuses specific to Google login are [here](https://developers.google.com/android/reference/com/google/android/gms/auth/api/signin/GoogleSignInStatusCodes) and all common status codes are [here](https://developers.google.com/android/reference/com/google/android/gms/common/api/CommonStatusCodes).  They suggest the following for Google login error handling:
<ul>
<li><a href="https://developers.google.com/android/reference/com/google/android/gms/common/api/CommonStatusCodes#INTERNAL_ERROR" rel="nofollow">INTERNAL_ERROR</a></li>
<li><a href="https://developers.google.com/android/reference/com/google/android/gms/common/api/CommonStatusCodes#INVALID_ACCOUNT" rel="nofollow">INVALID_ACCOUNT</a></li>
<li><a href="https://developers.google.com/android/reference/com/google/android/gms/common/api/CommonStatusCodes#NETWORK_ERROR" rel="nofollow">NETWORK_ERROR</a></li>
<li><a href="https://developers.google.com/android/reference/com/google/android/gms/auth/api/signin/GoogleSignInStatusCodes#SIGN_IN_CANCELLED" rel="nofollow">SIGN_IN_CANCELLED</a></li>
<li><a href="https://developers.google.com/android/reference/com/google/android/gms/auth/api/signin/GoogleSignInStatusCodes#SIGN_IN_FAILED" rel="nofollow">SIGN_IN_FAILED</a></li>
<li><a href="https://developers.google.com/android/reference/com/google/android/gms/common/api/CommonStatusCodes#SIGN_IN_REQUIRED" rel="nofollow">SIGN_IN_REQUIRED</a></li>
</ul>

While these errors can definitely happen, they are difficult to reproduce since they occur within the Google API calls.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java [hunk 1]
```diff
@@ -416,9 +427,46 @@ public void onActivityResult(int request, int result, Intent data) {
...
+                                showErrorDialog(getString(R.string.error_generic_network));
+                                break;
+                            // Cancelled by the user.
+                            case GoogleSignInStatusCodes.SIGN_IN_CANCELLED:
+                                AppLog.e(T.NUX, "Google Sign-in Failed: cancelled by user.");
+                                break;
+                            // Attempt didn't succeed with the current account.
+                            case GoogleSignInStatusCodes.SIGN_IN_FAILED:
+                                AppLog.e(T.NUX, "Google Sign-in Failed: current account failed.");
+                            // Attempted to connect, but the user is not signed in.
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: Is a `break` missing here or the fall through is intentional? If intentional, not sure if the second AppLog message that will be logged is compatible with this case.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java [hunk 2]
```diff
@@ -416,9 +427,46 @@ public void onActivityResult(int request, int result, Intent data) {
...
+                            case GoogleSignInStatusCodes.SIGN_IN_REQUIRED:
+                                AppLog.e(T.NUX, "Google Sign-in Failed: user is not signed in.");
+                            // Unknown error.
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: I'm assuming that the `break` is missing on purpose here. If so, let's add a comment that the fall through is intentional. Thanks!

File: WordPress/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -1899,5 +1900,10 @@
     <string name="login_epilogue_mysites_other">My sites</string>
     <string name="login_google_button_prefix">Or you can</string>
     <string name="login_google_button_suffix">Log in with Google.</string>
+    <string name="login_error_button">Close</string>
+    <string name="login_error_email_not_found">The Google account \'%s\' doesn\'t match any existing account on WordPress.com.</string>
+    <string name="login_error_generic">There was some trouble connecting with the Google account.</string>
+    <string name="login_error_suffix_internal">\nMaybe try again?</string>
```
Comment (hypest) [inline WordPress/src/main/res/values/strings.xml]: Not sure a question like that is very clear for the user. Better be more explicit.

But, if we're actually not sure ourselves whether the user should try again later or not, let's omit suggesting it.

File: WordPress/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -1899,5 +1900,10 @@
...
+    <string name="login_error_suffix_invalid">\nMaybe try a different account?</string>
```
Comment (hypest) [inline WordPress/src/main/res/values/strings.xml]: Same as above.

File: WordPress/src/main/res/values/strings.xml (hunk 3 above)
Comment (hypest) [reply WordPress/src/main/res/values/strings.xml]: @folletto I understand that that copy is taken from the design iteration. Would you mind if we rephrase them to non question form?

Comment (folletto) [reply WordPress/src/main/res/values/strings.xml]: It's a bit abstract to me to reconnect that to the screen where it appears, so can I get a screenshot to evaluate it? Context is relevant to see if it works.

Comment (hypest) [reply WordPress/src/main/res/values/strings.xml]: This is one of the cases that are not easy to replicate so, image this popup https://user-images.githubusercontent.com/3827611/29892211-07a06352-8d8b-11e7-8632-ab5f9b19f9f0.png
but having the text:
> There was some trouble connecting
> with the Google account.
> Maybe try again?
> 
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[CLOSE]

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java (hunk 1 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: The fall through was intentional, but the log was not.  I added the `break` in 0ea92a219ee918a8369285b0fc314c9bc8ab431c.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java (hunk 2 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: I added the `break` in 0ea92a219ee918a8369285b0fc314c9bc8ab431c to be more explicit.

File: WordPress/src/main/res/values/strings.xml (hunk 3 above)
Comment (folletto) [reply WordPress/src/main/res/values/strings.xml]: That's an excellent alternative to a screenshot! 🌟 

I'm ok in keeping the question there, but if you feel it's off we can either ping Editorial or remove that line. Both are ok for me in this scenario (reason: there's no clear reason forward and it's a modal with limited ability to add more context).

Comment (hypest) [reply WordPress/src/main/res/values/strings.xml]: Cool, let's remove it. @theck13 would that work OK with you too?

Comment (theck13) [reply WordPress/src/main/res/values/strings.xml]: To be clear, should we remove `login_error_suffix_internal`, `login_error_suffix_invalid`, or both?

***login_error_suffix_internal***
An internal Google API error occurred.  We don't know what happened.
> There was some trouble connecting
> with the Google account.
> Maybe try again?

***login_error_suffix_invalid***
The user attempted to connect with an invalid account name specified.  I'm not sure how a user would do this, but it's recommended by Google to check for it.
> There was some trouble connecting
> with the Google account.
> Maybe try a different account?

Comment (folletto) [reply WordPress/src/main/res/values/strings.xml]: The second seems a good positive hint for the user to try. These two work well as they are "symmetric", but the "Maybe try again?" is generic, and probably it's a kind of symmetry nobody would ever notice... 

tl;dr: remove "Maybe try again?" — keep "Maybe try a different account?"

Comment (theck13) [reply WordPress/src/main/res/values/strings.xml]: I think that's best.  I removed `login_error_suffix_internal`  in d042ff0f0d3dc2639cd1377de6b327b247816eb6.

File: WordPress/src/main/res/values/strings.xml (hunk 4 above)
Comment (hypest) [reply WordPress/src/main/res/values/strings.xml]: Since [we decided to keep this one](https://github.com/wordpress-mobile/WordPress-Android/pull/6674#discussion_r142998432), let's lose the question _mark_ and rephrasing this to "Please try a different account". Feel free to ping Editorial for better alternatives but in any case, this Alert is informational so no question is meant to be asked in the copy. Thanks!

Comment (hypest) [reply WordPress/src/main/res/values/strings.xml]: Let's change that in a future PR to lift the block here.

**Comment by hypest** [comment]
LGTM!
